lib/upgrade/etcd/conversionv1v3/bgppeer.go

```diff
@@ -0,0 +1,131 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package conversionv1v3
+
+import (
+	"fmt"
+	"net"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	apiv1 "github.com/projectcalico/libcalico-go/lib/apis/v1"
+	"github.com/projectcalico/libcalico-go/lib/apis/v1/unversioned"
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+	"github.com/projectcalico/libcalico-go/lib/scope"
+)
+
+// BGPPeer implements the Converter interface.
+type BGPPeer struct{}
+
+// APIV1ToBackendV1 converts v1 BGPPeer API to v1 BGPPeer KVPair.
+func (bp BGPPeer) APIV1ToBackendV1(rIn unversioned.Resource) (*model.KVPair, error) {
+	ap := rIn.(apiv1.BGPPeer)
+
+	if len(ap.Metadata.PeerIP.IP) == 0 {
+		return nil, fmt.Errorf("no PeerIP is set, invalid BGPPeer: %v", rIn)
+	}
+
+	k, err := bp.convertMetadataToKey(ap.Metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	d := model.KVPair{
+		Key: k,
+		Value: &model.BGPPeer{
+			PeerIP: ap.Metadata.PeerIP,
+			ASNum:  ap.Spec.ASNumber,
+		},
+	}
+
+	return &d, nil
+}
+
+func (_ BGPPeer) convertMetadataToKey(rmIn unversioned.ResourceMetadata) (model.Key, error) {
+	pm := rmIn.(apiv1.BGPPeerMetadata)
+
+	if pm.Scope == scope.Global {
+		if pm.Node != "" {
+			return nil, fmt.Errorf("With Global scope having a Node is invalid: %v", rmIn)
+		}
+		return model.GlobalBGPPeerKey{
+			PeerIP: pm.PeerIP,
+		}, nil
+	} else if pm.Scope == scope.Node {
+		if pm.Node == "" {
+			return nil, fmt.Errorf("With Node scope a Node must be defined: %v", rmIn)
+		}
+		return model.NodeBGPPeerKey{
+			PeerIP:   pm.PeerIP,
+			Nodename: pm.Node,
+		}, nil
+	} else {
+		return nil, errors.ErrorInsufficientIdentifiers{
+			Name: "scope",
+		}
+	}
+}
+
+// BackendV1ToAPIV3 converts v1 BGPPeer KVPair to v3 API.
+func (bp BGPPeer) BackendV1ToAPIV3(kvp *model.KVPair) (Resource, error) {
+	peer, ok := kvp.Value.(*model.BGPPeer)
+	if !ok {
+		return nil, fmt.Errorf("value is not a valid BGPPeer resource Value: %v", kvp.Value)
+	}
+
+	r := &apiv3.BGPPeer{
+		Spec: apiv3.BGPPeerSpec{
+			PeerIP:   peer.PeerIP.String(),
+			ASNumber: peer.ASNum,
+		},
+	}
+
+	switch kvp.Key.(type) {
+	case model.GlobalBGPPeerKey:
+		r.ObjectMeta = v1.ObjectMeta{Name: bp.ipToName(peer.PeerIP.IP)}
+	case model.NodeBGPPeerKey:
+		nk := kvp.Key.(model.NodeBGPPeerKey)
+		n := convertName(nk.Nodename)
+		r.Spec.Node = n
+		r.ObjectMeta = v1.ObjectMeta{Name: n + "." + bp.ipToName(peer.PeerIP.IP)}
```

Comment:
we had `Scope` as one of the keys in v1, so each resource would have `Scope` as part of its unique key. I'm wondering if ignoring the `Scope` in the name could cause a collision

Reply:
Should be fine - if the scope wasn't node then we won't include the node in the name (which is separated by a "." rather than the "-" separator used for the converted IP address.